dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,359 @@
+            self._affine = dwi_file.affine.copy()
+            return
+        if hasattr(value, 'affine'):
+            self._affine = value.affine
+        self._affine = np.array(value)
+
+    @gradients.setter
+    def gradients(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(value, skiprows=1)
+        self._gradients = value
+
+    @bvecs.setter
+    def bvecs(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(str(value)).T
+        else:
+            value = np.array(value, dtype='float32')
+
+        # Correct any b0's in bvecs misstated as 10's.
+        value[np.any(abs(value) >= 10, axis=1)] = np.zeros(3)
+        if self.bvals is not None and value.shape[0] != self.bvals.shape[0]:
+            raise ValueError('The number of b-vectors and b-values do not match')
+        self._bvecs = value
+
+    @bvals.setter
+    def bvals(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(str(value)).flatten()
+        if self.bvecs is not None and value.shape[0] != self.bvecs.shape[0]:
+            raise ValueError('The number of b-vectors and b-values do not match')
+        self._bvals = np.array(value)
+
+    def b0mask(self):
+        """Get a mask of low-b frames."""
+        return np.squeeze(self.gradients[..., -1] > self._b0_thres)
+
+    def normalize(self):
+        """Normalize (l2-norm) b-vectors."""
+        if self._normalized:
+            return
+
+        self._bvecs, self._bvals = normalize_gradients(
+            self.bvecs, self.bvals,
+            b0_threshold=self._b0_thres,
+            bvec_norm_epsilon=self._bvec_norm_epsilon,
+            b_scale=self._b_scale)
+        self._normalized = True
+
+    def generate_rasb(self):
+        """Compose RAS+B gradient table."""
+        if self.gradients is None:
+            self.normalize()
+            _ras = bvecs2ras(self.affine, self.bvecs)
+            self.gradients = np.hstack((_ras, self.bvals[..., np.newaxis]))
+
+    def generate_vecval(self):
+        """Compose a bvec/bval pair in image coordinates."""
+        if self.bvecs is not None:
+            self._bvecs = bvecs2ras(np.linalg.inv(self.affine), self.gradients[..., :-1])
+            self._bvals = self.gradients[..., -1].flatten()
+
+    @property
+    def pole(self):
+        """
+        Check whether the b-vectors cover a full or just half a shell.
+
+        If pole is all-zeros then the b-vectors cover a full sphere.
+
+        """
+        self.generate_rasb()
+        return calculate_pole(self.gradients[..., :-1], bvec_norm_epsilon=self._bvec_norm_epsilon)
+
+    def to_filename(self, path=None, bvecs=None, bvals=None):
+        """Write files (RASB, bvecs/bvals) to a given path."""
+        if path:
+            np.savetxt(str(path), self.gradients,
+                       delimiter='\t', header='\t'.join('RASB'),
+                       fmt=['%.8f'] * 3 + ['%g'])
+        if bvecs:
```

Comment:
Do we need to include a conditional here or does the shape get enforced by the interface? hard to tell
```
            if bvecs.shape[-1] == 3:
                np.savetxt(str(bvecs), self.bvecs.T, fmt='%.6f')
            else:
                np.savetxt(str(bvecs), self.bvecs, fmt='%.6f')
```

Reply:
At this point, `self.bvecs` has (Nx3) dimensions. Please note that `bvecs` here just contains a string or a Path (perhaps improving the docstring here would be nice).

Reply:
Oh, I see that now. I wonder if to avoid confusion, we should rename bvecs to fbvecs for every place we're referring to a file string/Path, and go with bvecs any time we're referring to an array?

Reply:
The function is called `to_filename`, I don't think changing the name of the arguments will change much. I guess it is more of an interface problem, i.e., should we have `to_filename` store only RASB and then add two more methods like `bvecs_to_filename` and `bvals_to_filename` or a `write_bvec_bval(path)` that adds `.bval` and `.bvec`.

Another alternative is:

``to_filename(filename, filetype='rasb')`` and then switch to bvec/bval mode if filetype is `'fsl'` I would probably like this last one better.

Reply:
Fair enough. Well I like the latter alternative best, but yes I think we need something like a write_bvec_bval(path) here. I sort of had something like this previously in the interface with:
```
if self.inputs.save_fsl_style is True:
    fbval_out_file, fbvec_out_file = vt.save_vecs_fsl()
```
but the to_filename would be best

Reply:
Implemented the `to_filename(filename, filetype='fsl')` route. :+1: